Handling supplementary (sub) alignments during haplotagging

aganezov opened the pull request:
As per updated doc, the following description provides the rational for the pull request and added logic to the `whatshap haplotag` command:

-----
When working with long or ultra-long reads (e.g., nanopore reads) resulting alignment may contain
both primary and supplementary alignments. Depending on the sample which reads are being haplotagged
it is important to recognize that sometimes the data used for variants phasing (into Phase blocks and haplotypes
within phase blocks). An example of such case can be working with Tumor / Normal cancer genomics.
A matching normal sample N can be used for germline (heterozygous) SNP detection and phasing,
producing the N.phased.vcf.gz file. Since tumor genome "inherits" all germline SNPs from the matching normal
(loss is possible with LOH events), it becomes important to haplotag tumor reads based on germline SNPs
and their phaseblocks/haplotype groupings.
Because cancer genomes can be highly rearranged with somatic (i.e., non-germline/non-matching-normal)
structural variations (e.g., translocations, long-range deletions/inversions) the reads that come from
derived (rearranged) chromosomes can have both primary and supplementary alignments that can fall into rather
distant regions of the genomes (even distinct chromosomes in case of translocations/chromoplexy/chromothripsis).
Such primary/supplementary alignment regions may fall into phased regions (based on N.phased.vcf.gz info)
with distinct phase block ids. In such cases it may beneficial to attempt to assign each alignment segment into
respective phase block/haplotype based on spanned phased variants. This can help identify haplotype/ps groupings of
somatic variants.

Depending on the needs of ones setup, the following strategies for treating supplementary alignments of a single read
using the ``--tag-supplementary`` CLI flag (and arguments for it enumerated below) and its values:

* ``skip`` / no flag specified -- each supplementary alignment segment is skipped in assigning PS/HP SAM tags
* ``copy-primary`` / flag without argument -- each supplementary alignment segment located in the same chromosome as the primary alignment segment get a copy of PS/HP SAM tags obtained for the primary alignment, if any obtained in haplotagging.
* ``independent-or-skip`` -- each supplementary alignment is treated as an independent (i.e. as are primary alignments) case and and attempt is made to haploptag it based on spanned phased variants. If unsuccessful, no PS/HP assignment is made.
* ``independent-or-copy-primary`` -- each supplementary alignment is treated as an independent (i.e. as are primary alignments) case and and attempt is made to haploptag it based on spanned phased variants. If unsuccessful, get a copy of PS/HP SAM tags obtained for the primary alignment, if any obtained in haplotagging and alignment segments located on the same chromosome.
-----

The added logic is added while respective backwards compatibility. CLI invocation allows both no flag usage, flag with no argument, as well as a flag + argument setups. 

An couple of examples of the cases where such more sophisticated handling of supplementary alignment segments is demonstrate below based on examples from a recently released public ONT dataset (https://www.ncbi.nlm.nih.gov/sra/?term=PRJNA1086849) of several Tumor + Normal cell line sample pairs. For example we've taken B1437 normal and H1437. B1437 was ran through the wf-human-variation Nanopore epi2me workflow (https://github.com/epi2me-labs/wf-human-variation) and the phased germline VCF was generated as a result. 

H1437 is a tumor cell line matching BL1437 normal one. It contains multiple somatic structural variations, identified with Severus. Below we show IGV screenshots of 2 somatic Structural variants. 

Top alignment track is the matching normal BL1437, second alignment track if H1437 haplotagged with current version of whatshap with `--tag-supplementary` and the bottom track is the same H1437 original alignment haplotagged with the pul request version of the whatshap using `--tag-supplementary independent-or-skip` strategy. 
Colors correspond to distinct PS block assignments. Grouping is either based on PS or HP tags, or supplementary tags respectively. 
We also show example read/alignment and differences in their PS/HP assignments as an example.

First is a translocation-like event involving chr8 and chr12. 
Region chr12 involved a lot of supplementary alignments most of which were not haplotagged in the latest version of whatshap as even with `--tag-supplementary` enabled would never copy PS/HP tags from primary alignments onto a supplementary ones.

![tr_example_1_1](https://github.com/whatshap/whatshap/assets/1204593/9d9c8957-9167-4f0c-94c2-dda41a3fc00b)
![tr_example_1_2](https://github.com/whatshap/whatshap/assets/1204593/cca83007-5c9c-4a42-a0d5-0c77a93f65e1)
![tr_example_2_1](https://github.com/whatshap/whatshap/assets/1204593/fa2cb2ba-d129-4cbf-b056-c5a2f3c52262)
![tr_example_2_2](https://github.com/whatshap/whatshap/assets/1204593/edf93c9f-e2f6-4813-bf10-a8ceb8c71f59)

Second is an inversion-like event on chromosome chr1.
Right side of the inversion:
![inv_example_1_1](https://github.com/whatshap/whatshap/assets/1204593/dc582207-75f8-48a0-b03d-e3178c112ebd)
![inv_example_1_2](https://github.com/whatshap/whatshap/assets/1204593/4db2b7ac-13f1-4739-9f5c-32a8925fed5f)
![inv_example_1_3](https://github.com/whatshap/whatshap/assets/1204593/514581dc-ea49-4fd4-9ec3-358a598be817)

Left side of the inversion:
![inv_example_2_1](https://github.com/whatshap/whatshap/assets/1204593/43640939-7e41-4477-a591-52287843706d)
![inv_example_2_2](https://github.com/whatshap/whatshap/assets/1204593/fa1cc4d2-ee89-49fb-aa7c-8d4048887e19)
![inv_example_2_3](https://github.com/whatshap/whatshap/assets/1204593/ff90ed8a-39b2-4168-b79c-3401a43d563b)

Inversion breakends fall into distinct phase blocks, thus the current strategy of "copy primary alignment PS/HP values onto supplementary alignments shows it limits with a mix of the PS/HP values being shown above.
It is important to note that the HP=2 assignment for both phase blocks in this case is a mere coincidence and HP assignments in either one of the phase blocks could have been flipped during phasing, resulting in a confusing HP values for inversion-supporting reads around both breakends.

aganezov committed:
adding target alignment chromosome, subalignmentid unique id for a read subalignment segment to distinguish it from other subalignments) and is_supplementary boolean flag to core Read class and exposing it into python ecosystem.
New fields are added as optional and in such order as to ensure backward compatability with positional-only constructor invocation

aganezov committed:
adding support for a comprehensive haplotagging of alignment segments, whether primary / supplementary most applicable for cases where phasing of variants is performed by un-rearranged reads (e.g., matching-normal sample) but the haplotagging is performed on somatic SV-spanning reads where primary alignments may fall into PS_A while supplementary alignment(s) fall into PS_B, PS_C, etc phase blocks and require attribution to respective phase block ids and haplotypes.

backwards compatability is preserved both from cli invocation ux as well as via internal logic handling.

mikolmogorov commented:
Subscribing to this, much needed feature for cancer SV analysis! We are developing tools  that rely on read haplotagging to produce haplotype specific calls (https://github.com/KolmogorovLab/Severus).

aganezov committed:
adding distance threshold (default 100Kbp) and strand match (default on) as cli options to haplotagging command to be in line with the phasing logic where supplementary alignments may be considered. Tunable on both parameters to disable if needed.

aganezov committed:
Docs update for adding distance threshold and strand match as cli options to haplotagging command.

marcelm commented on whatshap/cli/haplotag.py:
```diff
@@ -57,9 +79,38 @@ def add_arguments(parser):
+             " segment and is assigned (if possible) to PS/HP based on spanned phased variants. "
+             " If no assignment based on variants if possible, defaults to copying primary alignmtn assignment, "
+             " if exists and on the same chromosome."
+             " [default: with no flag -- skip, default with flag and no value -- copy-primary] ")
```
Please use round parentheses for consistency with the other help texts. Also, please remove the space before the `"`.

aganezov replied:
fixed

marcelm commented on whatshap/cli/haplotag.py:
```diff
@@ -57,9 +79,38 @@ def add_arguments(parser):
+        choices=list(SupplementaryHaplotaggingStrategy),
+        default=SupplementaryHaplotaggingStrategy.SKIP, const=SupplementaryHaplotaggingStrategy.COPY_PRIMARY,
+        dest="supplementary_strategy",
+        help="Also tag supplementary alignments. "
```
This help text is too long.

aganezov replied:
fixed

marcelm commented on src/read.h:
```diff
@@ -64,6 +71,11 @@ class Read {
 	int sample_id;
 	int id;
 	int reference_start;
+	std::string chromosome;
+	std::string read_subAlignment_id;
+	bool is_supplementary;
+	int reference_end;
+	bool is_reverse;
```
I’m not so keen on adding all these attributes to the `Read` class. I’ll have to think about this more, but an alternative is to track this information separately in a separate data structure (and only in `haplotag.py`).

marcelm commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,578 @@ def test_haplotag_cli_parser(tmp_path):
     assert ps_count > 0
 
 
+def test_haplotag_cli_parser_supplementary_strategy_no_flag(tmp_path):
```
This is a lot of (duplicated) code that essentially only tests whether ArgumentParser was used correctly. Please at least merge this and the next two functions into one. You can test all the options by using `pytest.mark.parametrize`.

aganezov replied:
fixed

marcelm commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,578 @@ def test_haplotag_cli_parser(tmp_path):
+                 supplementary_distance_threshold=1_000_000)
+
+    a: pysam.AlignedSegment
+    for a in pysam.AlignmentFile(out_bam_copy_primary_strategy):
```
Please always ensure the file gets closed correctly by using a context manager `with pyam.AlignmentFile(...) as ...:`).

aganezov replied:
fixed

marcelm commented on whatshap/core.pyx:
```diff
@@ -49,20 +49,23 @@ cdef class NumericSampleIds:
 	def inverse_mapping(self):
 		"""Returns a dict mapping numeric ids to sample names."""
 		return {numeric_id:name for name, numeric_id in self.mapping.items()}
-	
```
Please avoid unrleated whitespace-only changes (fixing the trailing whitespace in this file should be a separate PR).

aganezov replied:
fixed

marcelm reviewed (commented):
Thanks for this! This is not a small PR, so I am a bit overwhelmed by the amount of code changes, and also the length of your description. We are very few people who work on WhatsHap for at most a couple of hours each month, which makes it a bit difficult to review.

So I think I need some help (and more time) before I can do a review.

- Can you please summarize what this PR does in a single sentence? Something that can be put into the changelog.
- I approved the CI run now and noticed that many CI checks are failing. You can run the tests locally using `tox` to avoid this.
- We have unfortunately not had any contributing guidelines before, but I took this opportunity to add such a document, see #561. Can you please read through it?
- I added a first couple of review comments.

aganezov committed:
Addressing pull request comments / raised issues:
1. whitespace only changes removed from core.pyx
2. flake8 passing
3. black passing
4. python 3.8 argparse compatability
5. tests refactoring with code duplication reduction
6. help text length reduction / parenthesis usage alignment
7. context manager usage for alignment files in tests

aganezov commented:
Shortened pull request description:
Expanding read haplotagging logic to allow the treatment of the supplementary alignment as independent entities (while retaining backwards compatibility with the earlier logic).

aganezov commented:
is there any update on this PR? please let me know if I can of any help in way.

marcelm commented:
I plan to review this next week. Please ping me by Thursday if I haven’t gotten back to you by then. (You could resolve the conflicts in the meantime.)

aganezov committed:
Merge branch 'main' into supplementary_haplotagging

aganezov commented:
this is the past-Thursday ping

marcelm reviewed (changes requested):
I’ve spent some time on this now and again, the sheer amount of code changes makes it hard to review.

Realistically, I’m not going to have time to bring this into the shape that I would like it to be in, so I’m willing to live with some of the problems that I perceive, in particular when it comes to unimportant things like code style. I don’t want the limited time that I can spend on WhatsHap nowadays to stand in the way of this improvement being merged.

However, I am worried at the moment that the changes will negatively affect performance or memory usage when running `whatshap phase` because you made changes to the `Read` object, which is used in many WhatsHap subcommands.

So I’m willing to merge this 1) if you can somehow address my worry about increased memory usage by either reducing the memory usage or by showing that it’s not a problem, 2) by showing that there’s no slowdown in `phase`, 3) and by addressing the review comments that I make.

aganezov committed:
Addressing pull request comments / raised issues:
1. added subsection in docs
2. shortened help text in haplotag cli
3. updated code style for Read in read.cpp
4. removed argp shortened argparse import and made a single import in test_run_haplotag.py
5. underscore to single-letter variable fix in test_run_haplotag.py
6. removed description and prog from new tests in test_run_haplotag.py
7. substituted str(outbam) to os.devnull in test_run_haplotag.py when command is ont ran
8. refactored supplementary strategy pytest parameterization in test_run_haplotag.py to simplify test of cli flags
9. split cross-combination of multiple cli arguments testing in test_run_haplotag.py into separate tests
10. removed comments from test_run_haplotag.py which only made sense when looking at the diff
11. updated tests in test_run_haplotag.py to not have X.has_tag asserts as the test would correctly fail without them, increasing code readability
12. reduced complexity of the code in test_run_haplotag.py by reducing indentation levels where applicable
13. updated comments in test_run_haplotag.py to better decribe the outline of newly added test for supplementary read haplotagging testing
14. substituted 100_000 to 100,000 in haplotag.py cli help text to increase readability
15. updated comments in haplotag.py for ignore_read method to accommodate the change in parameter name
16. rearranged bool fields in read.h to have better alignment of struct fields

aganezov commented on doc/guide.rst:
```diff
@@ -766,6 +766,35 @@ heterozygous variants.
 
 |
 
+When working with long or ultra-long reads (e.g., nanopore reads) resulting alignment may contain
```
done

aganezov commented on whatshap/cli/haplotag.py:
```diff
@@ -57,9 +83,31 @@ def add_arguments(parser):
+             "(iii) independent-or-skip -- each supplementary alignment treated as an independent alignment segment; "
+             "(iv) independent-or-copy-primary -- each supplementary alignment treated as an independent alignment "
+             "defaults to primary PS and HP tags;\n"
+             "(default: with no flag -- skip, default with flag and no value -- copy-primary)")
```
done

aganezov commented on src/read.cpp:
```diff
@@ -8,7 +8,12 @@
+ 			const std::string& chromosome, const std::string& read_subAlignment_id, bool is_supplementary, int reference_end, bool is_reverse) : name(name), mapqs(1, mapq), source_id(source_id), sample_id(sample_id), reference_start(reference_start),
+ 				chromosome(chromosome), read_subAlignment_id(read_subAlignment_id), is_supplementary(is_supplementary),
+ 				reference_end(reference_end), is_reverse(is_reverse),
+ 			 	BX_tag(BX_tag), HP_tag(HP_tag), PS_tag(PS_tag) {
```
done

aganezov commented on whatshap/cli/haplotag.py:
```diff
@@ -446,7 +633,7 @@ def ignore_read(alignment, tag_supplementary):
     - IF not tag_supplementary AND is_supplementary
```
done

aganezov commented on src/read.h:
```diff
@@ -64,6 +71,11 @@ class Read {
+	std::string read_subAlignment_id;
+	bool is_supplementary;
+	int reference_end;
+	bool is_reverse;
```
done

aganezov commented on whatshap/cli/haplotag.py:
```diff
@@ -57,9 +83,31 @@ def add_arguments(parser):
+        default=100_000,
+        help="Maximum distance between supplementary alignment record and "
+             "a primary one for the tag copying onto the supplementary to be attempted.\n"
+             "(default: 100_000)")
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+
+def test_haplotag_cli_parser_supplementary_distance_threshold(tmp_path):
+    from whatshap.cli.haplotag import add_arguments as haplotag_add_arguments
+    import argparse as argp
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+    parser = argp.ArgumentParser(description="haplotag_test_parser", prog="whatshap_pytest")
+    haplotag_add_arguments(parser)
+    haplotag_args = parser.parse_args(
+        [_ for _ in supplementary_strategy_cli_flag.split("=") if len(_) > 0]
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+    import argparse as argp
+
+    outbam = tmp_path / "output.bam"
+    parser = argp.ArgumentParser(description="haplotag_test_parser", prog="whatshap_pytest")
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+        + [
+            "--no-reference",
+            "--output",
+            str(outbam),
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+        assert (
+            haplotag_args.supplementary_strategy
+            == SupplementaryHaplotaggingStrategy.INDEPENDENT_OR_COPY_PRIMARY
+        )
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+"""
+
+
+def test_run_haplotag_supplementary_skip(tmp_path):
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+            elif a1.query_name == "R2" and not a1.is_supplementary:
+                assert a1.has_tag("HP")
+                assert a2.has_tag("HP")
+                assert a1.get_tag("HP") == a2.get_tag("HP")
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+        out_bam_explicit_skip
+    ) as source2:
+        for a1, a2 in zip(source1, source2):
+            assert a1.query_name == a2.query_name
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -350,8 +1037,10 @@ def test_haplotag_supplementary(tmp_path):
         alignment_file="tests/data/haplotag.supplementary.bam",
         output=outbam2,
-        tag_supplementary=True,
+        supplementary_strategy=SupplementaryHaplotaggingStrategy.COPY_PRIMARY,  # switch from boolean to enum
```
done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+
+
+"""
+The idea is cover the use case of having a vcf produced and/or phased with long reads in a matching normal sample
```
updated comment. done

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+        "--tag-supplementary=independent-or-copy-primary",
+    ],
+)
+@pytest.mark.parametrize(
```
split into separate tests. done

aganezov commented on whatshap/cli/haplotag.py:
```diff
@@ -106,19 +154,110 @@ def get_variant_information(variant_table: VariantTable, sample: str):
+            # this would have equaled to have an alignment.query_name being present in read_to_haplotype data
+            #   in original logic where only COPY was present
+            raise KeyError
     except KeyError:
```
`alignment_representation`  and `min_alignment_distance` are the external methods called within the code block, non of which can produced a KeyError. The code itself only uses a 2 element list with `pop` methods that are guaranteed to not throw a `KeyError`. The updated check for extracting `haplotype`, `quality`, and `phaseset` 
```
...
            if repr in read_to_haplotype:
                haplotype, quality, phaseset = read_to_haplotype[repr]
...
```
has an explicit check as to not throw a `KeyError`. 
The only way for a `KeyError` to emerge is via an explicit `throw` which happens if no representation is present (identical logic to the previous code). Comments clearly outline the logic in the code.

aganezov commented on tests/test_run_haplotag.py:
```diff
@@ -111,6 +111,690 @@ def test_haplotag_cli_parser(tmp_path):
+    with pysam.AlignmentFile(out_bam_independent_or_skip_strategy) as source:
+        for a in source:
+            if a.query_name == "R1":
+                # chr1_PS1_H1
```
this comment clarifies which (sub)alignments of which read are considered, with the comments before the test that use the designed test case of alignments clearly outline. e.g.,

```
...
read R1 is represented by the following string of reference segment (rc == reverse complement; sub == subregion)
chr1_PS1_H1 -> chr1_NPS_H1 -> chr1_PS2_H1 -> chr1_rcPS1_H2 -> chr2_PS1_H1 -> chr1_rcPS1_H1

chr1_PS1_H1         -- supplem, cigar: 3442M4D7991M49443S, flag: 2048
...
```

aganezov commented:
I've ran a couple of tests for phasing with (1) whatshap installed from bioconda (2.4 version) `release` and (2) compiled merge request version of whatshap `mr` on a 30x normal WGS ONT dataset (1437BL from severus somatic SV paper; SRA: https://www.ncbi.nlm.nih.gov/sra/SRX23913874[accn]) , with GRCh38 reference and SNPs inferred by Clair3.  Below are the results. 
First, `time -v` command results on a single run on the same system:

`release`
```
# Resource usage
Maximum memory usage: 2.056 GB
Time spent reading BAM/CRAM:                 3597.3 s
Time spent parsing VCF:                        97.8 s
Time spent selecting reads:                   512.5 s
Time spent phasing:                          10281.3 s
Time spent writing VCF:                       123.4 s
Time spent finding components:                 42.4 s
Time spent on rest:                           315.2 s
Total elapsed time:                          14969.9 s
        Command being timed: "whatshap phase --distrust-genotypes --ignore-read-groups --only-snvs --reference grch38.main_chrs.fna -o sample.wh_conda24.vcf.gz sample.vcf.gz sample.sort.bam"
        User time (seconds): 14873.82
        System time (seconds): 55.29
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 4:09:30
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2056176
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 26402955
        Voluntary context switches: 94613
        Involuntary context switches: 14558
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

`mr`
```
# Resource usage
Maximum memory usage: 2.086 GB
Time spent reading BAM/CRAM:                 2323.6 s
Time spent parsing VCF:                        78.9 s
Time spent selecting reads:                   564.9 s
Time spent phasing:                          11694.5 s
Time spent writing VCF:                        96.0 s
Time spent finding components:                 56.1 s
Time spent on rest:                           310.2 s
Total elapsed time:                          15124.2 s
        Command being timed: "whatshap phase --distrust-genotypes --ignore-read-groups --only-snvs --reference grch38.main_chrs.fna -o sample.wh_MR.vcf.gz sample.vcf.gz sample.e82_lsk.sort.bam"
        User time (seconds): 15013.45
        System time (seconds): 75.80
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 4:12:04
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2086256
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 25383665
        Voluntary context switches: 5539
        Involuntary context switches: 83797
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

Furthermore, I've ran snakemake-based cluster submitted runs of the same 2 versions of whatshap on the same dataset, with the same cli arguments with 5 runs with 3 iteration each. Below are the whatshap benchmarking results:

`release`
```
==> whatshap_release.0.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
18636.3761      5:10:36 1842.87 1899.31 1836.98 1837.98 0.00    0.00    98.90   18432.08
18348.0057      5:05:48 2004.16 2060.72 1998.37 1999.44 0.00    0.00    99.59   18273.55
18395.0404      5:06:35 1831.68 1889.07 1825.66 1826.74 0.00    0.00    99.60   18321.81

==> whatshap_release.1.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
18650.7027      5:10:50 1830.77 1888.11 1824.70 1825.74 0.00    0.00    98.97   18458.42
18182.0797      5:03:02 2006.21 2063.80 2000.42 2001.49 0.00    0.00    99.92   18167.12
18230.5879      5:03:50 1841.95 1898.52 1836.14 1837.21 0.00    0.00    99.74   18182.74

==> whatshap_release.2.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
14997.7245      4:09:57 1836.23 1893.81 1818.89 1825.33 0.00    0.00    99.21   14879.06
14927.3530      4:08:47 1828.57 1885.15 1811.29 1817.70 0.00    0.00    99.48   14849.71
14571.0495      4:02:51 1844.11 1900.97 1826.99 1833.40 0.00    0.00    99.55   14505.16

==> whatshap_release.3.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
14128.8971      3:55:28 1839.78 1895.92 1833.53 1834.22 0.00    0.00    98.95   13980.12
13938.1147      3:52:18 1837.37 1893.82 1831.44 1832.13 0.00    0.00    99.84   13914.75
14068.8041      3:54:28 1830.13 1887.66 1824.24 1824.91 0.00    0.00    99.48   13995.69

==> whatshap_release.4.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
15039.9066      4:10:39 1879.49 2065.86 1832.31 1838.75 0.00    0.00    98.81   14861.36
14908.0383      4:08:28 1836.26 1892.96 1846.30 1852.70 0.00    0.00    99.21   14791.43
14758.4495      4:05:58 1836.41 1893.96 1819.06 1825.51 0.00    0.00    99.44   14675.53
```

`mr`
```
==> whatshap_mr.0.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
14009.3929      3:53:29 1852.77 1907.92 1837.39 1842.73 0.00    0.00    99.54   13945.45
14028.2406      3:53:48 1865.65 1919.70 1850.27 1855.59 0.00    0.00    99.74   13992.65
14266.4640      3:57:46 1927.70 2005.21 1895.30 1900.61 0.00    0.00    99.83   14243.61

==> whatshap_mr.1.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
19079.7293      5:17:59 1893.61 1956.77 1861.55 1862.57 0.00    0.00    99.48   18980.91
19065.5086      5:17:45 1852.41 1906.68 1846.47 1847.54 0.00    0.00    99.69   19005.53
19095.4436      5:18:15 1865.47 1920.86 1858.15 1859.22 0.00    0.00    99.75   19046.81

==> whatshap_mr.2.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
19080.7478      5:18:00 1861.66 1916.05 1855.87 1856.90 0.00    0.00    99.83   19047.23
19078.5613      5:17:58 1909.26 1964.57 1903.51 1904.58 0.00    0.00    99.87   19052.59
19072.2283      5:17:52 1869.95 1924.95 1863.66 1864.78 0.00    0.00    99.73   19019.37

==> whatshap_mr.3.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
14301.5333      3:58:21 1866.38 1922.98 1851.11 1856.45 0.00    0.00    99.60   14243.52
14178.6167      3:56:18 1854.59 1909.63 1839.28 1844.57 0.00    0.00    99.88   14161.35
13910.6679      3:51:50 1861.69 1917.14 1846.61 1851.92 0.00    0.00    99.78   13879.08

==> whatshap_mr.4.vcf.gz.log <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
14592.8011      4:03:12 1856.68 1913.17 1850.67 1851.34 0.00    0.00    99.64   14539.22
14599.5254      4:03:19 1855.56 1911.09 1849.80 1850.45 0.00    0.00    99.77   14565.55
14905.8118      4:08:25 1865.55 1919.95 1859.76 1860.44 0.00    0.00    99.74   14865.28
```

while there are some time fluctuations (which I attribute to system load differences on HPC nodes used in submission), I believe the overall numbers show that there is no significant memory/runtime differences between `release` and `mr` versions that would warrant code changes. 

If you believe there are other whatshap commands that may be affected by the proposed changes (in terms of runtime/memory usage) in a negative fashion -- please advise, I'd be happy to run more tests.

aganezov commented on src/read.h:
```diff
@@ -64,6 +71,11 @@ class Read {
 	int sample_id;
 	int id;
 	int reference_start;
+	std::string chromosome;
+	std::string read_subAlignment_id;
```
see comment below on runtime/memory usage

marcelm reviewed (commented):
Thanks a lot for addressing my review comments and also for the benchmark results! The variance is quite high, so we cannot say that there is no slowdown, but I agree it is likely small enough to not be as noticable if it exists.

I have two more comments, but I think the rest is fine.

There’s a test failure that was fixed by PR #591, so I suggest that you merge main into this PR (you could also rebase, but I think there’s a risk of the review comments no longer being attached properly).

No promises, but I’ll try to be quicker in the next round.

aganezov committed:
Merge remote-tracking branch 'refs/remotes/upstream/HEAD' into supplementary_haplotagging

aganezov committed:
adjusting comments to not refer to old code and improving code flow/readability to ensure that non-haplotagged control flow happens only of all representation are exhausted with for/break/else rather than KeyError raising.

aganezov commented on whatshap/cli/haplotag.py:
```diff
@@ -106,19 +153,110 @@ def get_variant_information(variant_table: VariantTable, sample: str):
+            # we arrive here if no representation (however many we consider based on haplotagging strategy)
+            #   were present in info.
+            # this would have equaled to have an alignment.query_name being present in read_to_haplotype data
+            #   in original logic where only COPY was present
```
makes sense, updated and fixed.

marcelm commented:
Remember to fix the test failures (the linter complains)

aganezov committed:
black linter compliance

aganezov committed:
flake8 compliance

aganezov commented:
Here are the benchmarkign results for `phase` and `haplotag` commands between 2.4 conda install (`release`) and most recent update (`mr`). I've used both ONT ~30x dataset and Illumina ~40x dataset. Same set of SNPs was used for `phase` command. Same set of phased SNPs was used for `haplotag` command. The only difference was the bam file. 
The same command on the same inputs was run in 5 instances with 3 repeats each using a snakemake wrapper. 

ONT results. 
=========
Phasing.
---------

`release`:

```
==> whatshap_release.0.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
20684.1322      5:44:44 1844.71 1900.98 1827.03 1829.62 0.00    0.00    99.58   20597.08
19820.1409      5:30:20 1837.34 1893.70 1819.82 1823.10 0.00    0.00    99.77   19773.92
19848.2702      5:30:48 1842.82 1899.07 1825.07 1828.33 0.00    0.00    99.85   19818.77

==> whatshap_release.1.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
20413.2018      5:40:13 1844.23 1902.80 1826.88 1829.46 0.00    0.00    99.90   20392.20
19785.9876      5:29:45 1844.42 1901.94 1827.06 1830.32 0.00    0.00    99.84   19754.28
19535.0863      5:25:35 1844.71 1901.94 1827.03 1830.29 0.00    0.00    99.92   19519.64

==> whatshap_release.2.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
15567.0247      4:19:27 1846.09 1903.32 1828.42 1831.59 0.00    0.00    99.42   15476.18
15199.5008      4:13:19 1987.82 2045.12 1970.20 1974.21 0.00    0.00    99.78   15165.66
15214.4757      4:13:34 1829.98 1949.70 1812.17 1816.21 0.00    0.00    99.54   15142.95

==> whatshap_release.3.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
14569.0988      4:02:49 1923.44 2028.94 1848.80 1853.00 0.00    0.00    99.18   14448.44
13975.5748      3:52:55 1845.85 1902.96 1828.12 1834.54 0.00    0.00    99.82   13949.97
13901.0217      3:51:41 1840.54 1898.46 1822.58 1829.05 0.00    0.00    99.61   13845.09

==> whatshap_release.4.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
15577.7008      4:19:37 1830.21 1888.30 1812.39 1815.50 0.00    0.00    99.52   15502.52
15319.2406      4:15:19 1844.03 1901.11 1826.21 1830.24 0.00    0.00    99.44   15233.25
15310.5149      4:15:10 1841.72 1898.02 1824.16 1828.19 0.00    0.00    99.74   15269.46
```

`mr`

```
==> whatshap_mr.0.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
19941.8118      5:32:21 1889.91 1945.08 1874.72 1877.54 0.00    0.00    99.27   19794.65
18235.3366      5:03:55 1895.24 1978.79 1884.25 1889.56 0.00    0.00    99.74   18186.85
18917.6901      5:15:17 1902.23 1956.54 1887.02 1892.32 0.00    0.00    99.79   18875.98

==> whatshap_mr.1.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
19910.1540      5:31:50 1885.71 1940.03 1870.52 1873.29 0.00    0.00    99.66   19841.50
18910.6035      5:15:10 1894.23 1948.07 1878.61 1883.94 0.00    0.00    99.79   18870.04
18918.6106      5:15:18 1954.89 2013.71 1920.55 1925.87 0.00    0.00    99.84   18888.16

==> whatshap_mr.2.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
19715.7241      5:28:35 1891.73 1945.58 1876.08 1881.05 0.00    0.00    99.94   19704.39
19426.8262      5:23:46 1892.91 1947.63 1877.55 1882.43 0.00    0.00    99.76   19379.57
19298.1391      5:21:38 1889.72 1943.88 1874.38 1879.39 0.00    0.00    99.83   19265.89

==> whatshap_mr.3.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
15506.0114      4:18:26 1891.88 1946.37 1876.05 1877.23 0.00    0.00    99.89   15487.72
15223.7787      4:13:43 1894.73 1951.24 1879.75 1881.81 0.00    0.00    99.76   15185.55
15229.0080      4:13:49 1890.06 1947.46 1874.97 1877.02 0.00    0.00    99.84   15203.59

==> whatshap_mr.4.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
15547.8353      4:19:07 1898.63 1954.63 1883.09 1884.26 0.00    0.00    99.69   15499.45
15220.7664      4:13:40 1894.49 1949.42 1879.08 1881.22 0.00    0.00    99.80   15189.48
15208.7734      4:13:28 1895.94 1950.95 1880.57 1882.60 0.00    0.00    99.82   15180.02
```

haplotagging
-------------

`release`

```
==> whatshap_release.0.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
2277.5672       0:37:57 1022.03 1377.55 1005.00 1008.10 0.00    0.00    310.58  7074.07
2166.3243       0:36:06 937.41  1378.55 903.70  906.76  0.00    0.00    330.33  7155.98
2186.7687       0:36:26 849.98  1378.55 832.68  835.80  0.00    0.00    328.13  7175.43

==> whatshap_release.1.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
1608.4648       0:26:48 862.21  1481.22 838.54  840.27  0.00    0.00    305.91  4921.13
1639.6837       0:27:19 849.25  1378.54 831.86  833.58  0.00    0.00    310.48  5090.99
1627.5455       0:27:07 992.55  1454.81 998.47  1000.17 0.00    0.00    313.88  5108.50

==> whatshap_release.2.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
1704.9482       0:28:24 849.36  1377.54 831.77  833.52  0.00    0.00    295.15  5032.75
1626.7664       0:27:06 864.95  1377.54 838.43  840.13  0.00    0.00    311.76  5072.08
1594.9806       0:26:34 1021.87 1376.54 1005.05 1006.76 0.00    0.00    313.15  4995.11

==> whatshap_release.3.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
1669.8989       0:27:49 907.41  1377.54 885.52  887.23  0.00    0.00    312.09  5212.05
1620.8996       0:27:00 1021.61 1377.54 993.93  995.64  0.00    0.00    310.21  5028.71
1629.2915       0:27:09 849.35  1378.71 831.86  833.61  0.00    0.00    311.82  5080.81

==> whatshap_release.4.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
1661.9324       0:27:41 971.13  1379.54 943.70  945.42  0.00    0.00    306.35  5091.72
1594.6101       0:26:34 849.26  1377.54 831.77  833.50  0.00    0.00    316.45  5046.23
1657.6202       0:27:37 854.56  1377.54 832.31  834.01  0.00    0.00    310.09  5140.57
```

`mr`

```
==> whatshap_mr.0.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
1848.0519       0:30:48 903.98  1363.04 882.41  883.51  0.00    0.00    310.65  5741.43
1807.0027       0:30:07 977.42  1365.19 985.02  986.14  0.00    0.00    311.57  5630.05
1809.8274       0:30:09 1011.12 1364.19 996.70  997.79  0.00    0.00    316.02  5719.58

==> whatshap_mr.1.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
1917.2540       0:31:57 872.90  1364.19 851.16  852.28  0.00    0.00    305.22  5852.47
1849.0240       0:30:49 980.59  1415.78 954.86  955.98  0.00    0.00    307.02  5676.94
1800.8063       0:30:00 840.01  1365.19 824.79  825.92  0.00    0.00    318.47  5735.05

==> whatshap_mr.2.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
1925.1496       0:32:05 1011.24 1364.20 996.73  997.83  0.00    0.00    299.58  5768.19
1884.9414       0:31:24 916.93  1363.04 881.25  882.34  0.00    0.00    303.70  5725.13
1764.6888       0:29:24 958.03  1363.19 937.95  939.04  0.00    0.00    320.84  5662.01

==> whatshap_mr.3.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
1720.0617       0:28:40 838.45  1363.20 823.24  825.13  0.00    0.00    300.76  5173.73
1655.5311       0:27:35 943.79  1433.81 922.83  924.72  0.00    0.00    305.89  5064.24
1633.1645       0:27:13 1010.50 1363.19 995.98  997.85  0.00    0.00    305.50  4989.34

==> whatshap_mr.4.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
2293.0667       0:38:13 1011.43 1364.05 996.69  999.30  0.00    0.00    315.78  7241.67
2325.4795       0:38:45 1007.18 1363.21 985.20  987.77  0.00    0.00    311.52  7244.43
2283.6910       0:38:03 987.04  1389.23 964.34  966.89  0.00    0.00    317.47  7250.06
```

Illumina results. 
=========
Phasing.
---------

`release`

```
==> whatshap_release.0.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
19726.5592      5:28:46 8146.46 8240.74 8129.04 8131.60 0.00    0.00    99.64   19654.87
18850.0336      5:14:10 8146.35 8240.61 8128.89 8132.15 0.00    0.00    99.63   18779.83
19487.2770      5:24:47 8146.36 8239.73 8128.98 8132.24 0.00    0.00    99.67   19422.35

==> whatshap_release.1.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
18375.4946      5:06:15 8148.38 8240.62 8130.87 8133.46 0.00    0.00    99.63   18308.06
18647.1786      5:10:47 8148.08 8242.62 8130.84 8134.09 0.00    0.00    99.52   18557.22
18567.5289      5:09:27 8148.39 8241.81 8131.00 8134.26 0.00    0.00    99.62   18497.54

==> whatshap_release.2.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
19782.1146      5:29:42 8146.74 8219.61 8129.68 8133.58 0.00    0.00    99.64   19709.87
19128.2395      5:18:48 8146.43 8240.62 8140.39 8141.07 0.00    0.00    99.72   19073.27
17235.4485      4:47:15 8273.05 8396.55 8267.29 8267.95 0.00    0.00    99.68   17180.47

==> whatshap_release.3.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
14252.1069      3:57:32 8146.36 8239.60 8128.93 8133.12 0.00    0.00    99.63   14199.45
14627.0359      4:03:47 8146.42 8238.71 8129.00 8135.43 0.00    0.00    99.70   14582.70
13292.7022      3:41:32 8215.53 8311.09 8197.85 8204.33 0.00    0.00    99.53   13229.58

==> whatshap_release.4.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
14608.5459      4:03:28 8146.27 8240.71 8129.05 8132.15 0.00    0.00    99.73   14569.24
14302.0369      3:58:22 8313.85 8442.27 8296.59 8300.62 0.00    0.00    99.61   14245.83
14310.2089      3:58:30 8172.99 8271.68 8155.77 8159.80 0.00    0.00    99.51   14238.80
```


`mr`

```
==> whatshap_mr.0.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
20138.6107      5:35:38 9532.81 9644.34 9517.24 9518.94 0.00    0.00    99.70   20078.31
19516.2280      5:25:16 9495.45 9569.23 9480.48 9482.53 0.00    0.00    99.51   19420.27
19235.9787      5:20:35 9495.66 9570.24 9480.51 9484.16 0.00    0.00    99.63   19163.32

==> whatshap_mr.1.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
18675.0526      5:11:15 9494.39 9569.24 9479.88 9483.20 0.00    0.00    99.83   18642.32
18293.6672      5:04:53 9495.67 9588.25 9481.14 9484.50 0.00    0.00    99.74   18245.42
16978.4102      4:42:58 9493.79 9568.26 9488.85 9489.40 0.00    0.00    99.92   16964.57

==> whatshap_mr.2.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
20136.3400      5:35:36 9564.55 9680.75 9549.32 9551.02 0.00    0.00    99.31   19996.90
19640.7798      5:27:20 9493.66 9574.22 9478.49 9480.52 0.00    0.00    99.44   19529.78
18628.1338      5:10:28 9495.93 9576.22 9480.49 9484.17 0.00    0.00    99.49   18531.31

==> whatshap_mr.3.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
20275.5884      5:37:55 9495.52 9576.21 9480.48 9482.16 0.00    0.00    99.35   20142.96
19554.4583      5:25:54 9703.88 9817.79 9688.89 9690.94 0.00    0.00    99.47   19450.50
19578.5035      5:26:18 9493.58 9568.19 9478.45 9482.11 0.00    0.00    99.67   19512.96

==> whatshap_mr.4.vcf.gz.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
18573.1303      5:09:33 9493.77 9585.20 9479.09 9481.58 0.00    0.00    99.66   18509.41
18022.3477      5:00:22 9680.19 9789.62 9674.33 9675.04 0.00    0.00    99.77   17979.89
18455.7310      5:07:35 9493.68 9568.22 9488.86 9489.43 0.00    0.00    99.89   18434.94
```


haplotagging
-------------

`release`

```
==> whatshap_release.0.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
13084.8196      3:38:04 983.42  1346.71 977.99  978.96  0.00    0.00    151.66  19844.90
12550.6998      3:29:10 816.67  1347.71 810.85  811.80  0.00    0.00    150.06  18833.83
12691.0744      3:31:31 982.81  1277.66 976.99  977.97  0.00    0.00    147.56  18727.34

==> whatshap_release.1.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
8396.0358       2:19:56 815.64  1346.70 809.64  810.15  0.00    0.00    152.07  12768.15
8568.6740       2:22:48 905.98  1345.70 887.69  888.25  0.00    0.00    151.47  12979.23
8474.4159       2:21:14 931.55  1577.68 919.85  920.43  0.00    0.00    152.10  12889.25

==> whatshap_release.2.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
10456.5523      2:54:16 917.05  1281.22 886.20  892.67  0.00    0.00    163.32  17077.99
12364.7635      3:26:04 983.01  1277.37 966.67  973.11  0.00    0.00    150.97  18666.91
12938.2057      3:35:38 848.99  1279.18 818.39  824.86  0.00    0.00    149.92  19397.20

==> whatshap_release.3.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
13379.6252      3:42:59 814.73  1281.71 797.90  804.43  0.00    0.00    148.58  19879.59
11544.7365      3:12:24 960.14  1276.61 936.29  942.78  0.00    0.00    152.30  17583.04
10753.3285      2:59:13 816.64  1347.79 966.94  973.42  0.00    0.00    159.98  17202.81

==> whatshap_release.4.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
8582.7303       2:23:02 895.91  1345.70 885.35  886.03  0.00    0.00    157.50  13517.78
8435.0016       2:20:35 974.36  1282.70 956.78  957.44  0.00    0.00    158.70  13386.03
8708.4324       2:25:08 831.83  1345.70 813.44  814.13  0.00    0.00    156.65  13641.88
```


`mr`

```
==> whatshap_mr.0.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
11615.1606      3:13:35 970.97  1265.70 934.38  937.85  0.00    0.00    137.61  15983.84
11953.1506      3:19:13 862.87  1265.70 838.35  841.76  0.00    0.00    136.62  16330.40
11498.9773      3:11:38 904.59  1261.60 884.28  887.69  0.00    0.00    137.80  15845.38

==> whatshap_mr.1.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
16232.9148      4:30:32 972.51  1330.21 953.28  954.23  0.00    0.00    135.63  22017.09
15530.3716      4:18:50 843.57  1204.21 830.41  831.34  0.00    0.00    138.06  21441.26
16451.2358      4:34:11 964.45  1270.82 946.45  947.42  0.00    0.00    141.33  23250.19

==> whatshap_mr.2.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
12390.1788      3:26:30 804.33  1331.70 789.46  794.57  0.00    0.00    138.65  17178.39
12519.6626      3:28:39 953.02  1261.81 923.50  928.55  0.00    0.00    138.20  17302.03
12996.4294      3:36:36 971.88  1265.70 957.88  962.92  0.00    0.00    135.68  17633.83

==> whatshap_mr.3.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
11754.2758      3:15:54 963.15  1330.20 926.52  929.96  0.00    0.00    135.30  15903.08
11644.8870      3:14:04 829.80  1332.20 810.36  813.82  0.00    0.00    135.18  15741.86
11929.7274      3:18:49 827.46  1332.20 800.25  803.67  0.00    0.00    136.11  16236.60

==> whatshap_mr.4.bam.bench <==
s       h:m:s   max_rss max_vms max_uss max_pss io_in   io_out  mean_load       cpu_time
12639.3203      3:30:39 972.98  1236.09 946.65  951.69  0.00    0.00    138.11  17456.65
12450.6953      3:27:30 803.54  1267.20 788.55  793.67  0.00    0.00    138.27  17214.95
11483.3306      3:11:23 803.48  1269.24 786.77  792.13  0.00    0.00    138.72  15929.69
```


moderated increase in peak RAM usage for Illumina ~40x dataset in phasing instance (8->9.5GB). I would argue that on modern systems this is no a blocking issue. Other measurements support consistent time and memory usage performance (though variance is relatively high, but it can be attributed to different compute nodes on HPS system to which the respective commands were submitted for execution).

eesiribloom commented:
@aganezov out of interest, would the same scenarios you outlined with regards to supplementary alignments be present if you were phasing directly with a tumour bam instead of haplotagging from a matched normal?

aganezov commented:
> @aganezov out of interest, would the same scenarios you outlined with regards to supplementary alignments be present if you were phasing directly with a tumour bam instead of haplotagging from a matched normal?

@eesiribloom  all the issues/possibilities remain if you work with tumor only bam, rather than having a matching normal. If you can identify and phase SNPs into HP/PS tag groups, then haplotagging is indeed possible, and handling supplementary alignments independently is indeed beneficial, as one is interested in PS/HP read assignment based reference-localized alignments, not based on regardless where the primary lies. inversion-like example above demonstrates this clearly.

marcelm reviewed (approved):
Thanks so much for this contribution and also your patience! This is a feature that has been requested quite a bit, so I’m glad we’re finally adding it. I’ll merge this now and plan for a release soon.

mikolmogorov commented:
Thank you!

eesiribloom commented:
thanks @aganezov and others for putting this together! In rearranged cancer genomes is the `independent-or-skip` method recommended?

aganezov commented:
@eesiribloom that is correct, the `independent-or-skip` would be best suited for rearranged cancer genomes
